### PR TITLE
[LoadStoreVectorizer] Postprocess and merge equivalence classes

### PR DESCRIPTION
**Pull request by v-klochkov**
This patch introduces a new method:
void Vectorizer::mergeEquivalenceClasses(EquivalenceClassMap &EQClasses) const

The method is called at the end of Vectorizer::collectEquivalenceClasses() and is needed to merge equivalence classes that differ only by their underlying objects (UO1 and UO2), where UO1 is 1-level-indirection underlying base for UO2. This situation arises due to the limited lookup depth used during the search of underlying bases with llvm::getUnderlyingObject(ptr).

Using any fixed lookup depth can result into creation of multiple equivalence classes that only differ  by 1-level indirection bases.

The new approach merges equivalence classes if they have adjacent bases (1-level indirection). If a series of equivalence classes form ladder formed of 1-step/level indirections, they are all merged into a single equivalence class. This provides more opportunities for the load-store vectorizer to generate better vectors.

### CODE REVIEW

File: llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll
Comment (arsenm) [inline llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll:10]:
```suggestion
; workload; the offsets in getelementptr instructions are similar or same for
```

File: llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:1398]: use c++17 structure binding

File: llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:1402]: use c++17 structure binding

File: llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp
Comment (v-klochkov) [reply llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp]: Thank you for the comment! I made an incremental fix for it: https://github.com/llvm/llvm-project/pull/114501/commits/91a74c80e89ff46dcaf15dad30f9667e082d5979

File: llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp
Comment (v-klochkov) [reply llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp]: Fixed in https://github.com/llvm/llvm-project/pull/114501/commits/91a74c80e89ff46dcaf15dad30f9667e082d5979 . Thank you.

File: llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll
Comment (v-klochkov) [reply llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll]: the misprint is fixed: https://github.com/llvm/llvm-project/pull/114501/commits/91a74c80e89ff46dcaf15dad30f9667e082d5979

File: llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll
Comment (arsenm) [inline llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll:2]: Avoid the temporary file, nearly all tests should just pipe directly to FileCheck

File: llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll
Comment (arsenm) [inline llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll:21]: Why do some of the GEPs have inbounds and some not? If they are not significant, they should probably be removed

File: llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll
Comment (arsenm) [inline llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll:15]: Use update_test_checks so we can see full context

File: llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:1428]:
```suggestion
  LLVM_DEBUG({
    dbgs() << "LSV: mergeEquivalenceClasses: after merging:\n";
    for (const auto &EC : EQClasses) {
      dbgs() << "    Key: ([" << std::get<0>(EC.first)
             << "]: " << *std::get<0>(EC.first) << ", " << std::get<1>(EC.first)
             << ", " << std::get<2>(EC.first) << ", "
             << static_cast<int>(std::get<3>(EC.first)) << ")\n";
      for (const auto &Inst : EC.second)
        dbgs() << "\tInst: " << *Inst << '\n';
    }
  });
```

File: llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:1407]: No auto. Should these be references?

File: llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:1366]:
```suggestion
  LLVM_DEBUG({
    dbgs() << "LSV: mergeEquivalenceClasses: before merging:\n";
    for (const auto &EC : EQClasses) {
      dbgs() << "    Key: ([" << std::get<0>(EC.first)
             << "]: " << *std::get<0>(EC.first) << ", " << std::get<1>(EC.first)
             << ", " << std::get<2>(EC.first) << ", "
             << static_cast<int>(std::get<3>(EC.first)) << ")\n";
      for (const auto &Inst : EC.second)
        dbgs() << "\tInst: " << *Inst << '\n';
    }
  });
  LLVM_DEBUG({
    dbgs() << "LSV: mergeEquivalenceClasses: RedKeyToUOMap:\n";
    for (const auto &RedKeyToUO : RedKeyToUOMap) {
      dbgs() << "    Reduced key: (" << std::get<0>(RedKeyToUO.first) << ", "
             << std::get<1>(RedKeyToUO.first) << ", "
             << static_cast<int>(std::get<2>(RedKeyToUO.first)) << ") --> "
             << RedKeyToUO.second.size() << " underlying objects:\n";
      for (auto UObject : RedKeyToUO.second)
        dbgs() << "    [" << UObject << "]: " << *UObject << '\n';
    }
  });
```

File: llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:327]:
```suggestion
  /// Merges the equivalence classes if they have underlying objects that differ
```

File: llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp
Comment (arsenm) [inline llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:1377]:
```suggestion
          getUnderlyingObject(UObject, MaxLookupDepth);
```

**Comment by arsenm** [comment]
> Using any fixed lookup depth can result into creation of multiple equivalence classes that only differ by 1-level indirection bases.

The test IR doesn't look canonical / optimal. Other passes are expected to have rewritten the addressing expressions to be a simpler form. If I run your examples through -O3, it mostly vectorizes with some quirks:

1. have to hackily avoid a memset that forms
2. Ends up with a scalar store + a 7x vector store

File: llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll
Comment (v-klochkov) [reply llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll]: Fixed as suggested: https://github.com/llvm/llvm-project/pull/114501/commits/427983a9378f8612c054ad219023564ecf803643

File: llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll
Comment (v-klochkov) [reply llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll]: There were some in the original workload. This test is the result of big simplification.
Those `inbound` are not important for the test, so I removed them as suggested: https://github.com/llvm/llvm-project/pull/114501/commits/427983a9378f8612c054ad219023564ecf803643

File: llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll
Comment (v-klochkov) [reply llvm/test/Transforms/LoadStoreVectorizer/X86/massive_indirection.ll]: Ok, added more context using `update_test_checks.py`. Thank you.

File: llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp
Comment (v-klochkov) [reply llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp]: `const auto &` would be good here, IMO. Added it here: https://github.com/llvm/llvm-project/pull/114501/commits/427983a9378f8612c054ad219023564ecf803643

**Comment by v-klochkov** [comment]
> > Using any fixed lookup depth can result into creation of multiple equivalence classes that only differ by 1-level indirection bases.
> 
> The test IR doesn't look canonical / optimal. Other passes are expected to have rewritten the addressing expressions to be a simpler form. If I run your examples through -O3, it mostly vectorizes with some quirks:
> 
> 1. have to hackily avoid a memset that forms
> 2. Ends up with a scalar store + a 7x vector store

@arsenm : Matt - thank you for the review.
I added an incremental commit https://github.com/llvm/llvm-project/pull/114501/commits/427983a9378f8612c054ad219023564ecf803643 to address all of your comments.

Regarding the case itself. The original workload was way more complex, the LIT only shows the idea of it, it is a huge simplification.
For the 1st case/pattern I saw `7x` + `1x` vectors (for loads and stores), which ruined performance - `7x` vectors had to be lowered/legalized to `4x`+`2x`+`1x` producing 4 mem-ops + extra reg-to-reg moves and swizzles instead of a single `8x` mem-operation.

The 2nd pattern in LIT shows potentially worse situation, giving 8 equivalence classes having 1 scalar mem-operation in each.

**Review by michalpaszkowski — APPROVED** [review]
Thank you for the review @arsenm, please post any other comments/suggestions or let me know if anything wasn't addressed.

The PR looks good to me. @v-klochkov does not have write access, I can merge the PR early next week (waiting in case there are any other comments).

**Comment by arsenm** [comment]
> Regarding the case itself. The original workload was way more complex, the LIT only shows the idea of it, it is a huge simplification. For the 1st case/pattern I saw `7x` + `1x` vectors (for loads and stores), which ruined performance - `7x` vectors had to be lowered/legalized to `4x`+`2x`+`1x` producing 4 mem-ops + extra reg-to-reg moves and swizzles instead of a single `8x` mem-operation.
> 

Can you add a case that's more representative? You can try llvm-reduce to get a sample

**Comment by v-klochkov** [comment]
> > Regarding the case itself. The original workload was way more complex, the LIT only shows the idea of it, it is a huge simplification. For the 1st case/pattern I saw `7x` + `1x` vectors (for loads and stores), which ruined performance - `7x` vectors had to be lowered/legalized to `4x`+`2x`+`1x` producing 4 mem-ops + extra reg-to-reg moves and swizzles instead of a single `8x` mem-operation.
> 
> Can you add a case that's more representative? You can try llvm-reduce to get a sample

It may be challenging due to couple of reasons. There is a legal consideration: using llvm-reduce on non-disclosed code might not make it shareable. and second - the original input code is for a target that is not published yet, manually removing those target specifics details may result into pretty synthetic test again.

The existing LIT test is designed to be easily understandable. Adding more variations into the address-operation (use different offsets in getelementptr, or using slightly different operations - not only getelementptr), keeps the idea of the test the same.

Perhaps, we can both agree that the 6-steps-depth-limited search of the underlying object can easily lead to situations where equivalence classes have underlying objects that are only one step or level apart.

@arsenm - I am open to adding more variations in address operations if you believe it would improve the LIT test. Please let me know your thoughts first.

**Comment by v-klochkov** [comment]
> Can you add a case that's more representative? You can try llvm-reduce to get a sample

@arsenm - Matt - Please take a look at 1 extra LIT case here: https://github.com/llvm/llvm-project/pull/114501/commits/7b07f8cf1e9ad1345bcc97fe9c99d1bb7590d819

llvm-reduce is a great tool, I played with it; In this case though, it turned out to be easier to copy the interesting STOREs operations and minimally re-build missing IRs around them to show the idea of the context the STOREs in the original workload were placed into.
It is easy to get underlying objects that 1 indirection level apart from each other when the 1st store in a sequence is a store indexed by `index` and consequent stores indexed by `index` + `some_const`, which gives extra depth/level created by getelementptr `index + some_const`.

**Comment by v-klochkov** [comment]
> > Can you add a case that's more representative? You can try llvm-reduce to get a sample
> 
> @arsenm - Matt - Please take a look at 1 extra LIT case here: [7b07f8c](https://github.com/llvm/llvm-project/commit/7b07f8cf1e9ad1345bcc97fe9c99d1bb7590d819)
> 
> llvm-reduce is a great tool, I played with it; In this case though, it turned out to be easier to copy the interesting STOREs operations and minimally re-build missing IRs around them to show the idea of the context the STOREs in the original workload were placed into. It is easy to get underlying objects that 1 indirection level apart from each other when the 1st store in a sequence is a store indexed by `index` and consequent stores indexed by `index` + `some_const`, which gives extra depth/level created by getelementptr `index + some_const`.

@arsenm - Greetings, this is just a reminder/ping message. Thank you!

**Review by michalpaszkowski — CHANGES_REQUESTED** [review]
@v-klochkov Thank you for adding the test and addressing the comments! I just ran the code formatter. One more thing, could you please name the unnamed/numbered vregs in the test -- this should help with any changes in the future? Please rebase on top of `main` branch -- once the builbots pass, I will merge the change.

**Comment by v-klochkov** [comment]
> I just ran the code formatter. One more thing, could you please name the unnamed/numbered vregs in the test -- this should help with any changes in the future? Please rebase on top of `main` branch -- once the builbots pass, I will merge the change.

@michalpaszkowski, thank you! I re-based the patch, removed unnamed/numbered vregs, removed `undef`: 4bff74970fc7d80dd5d261770d026c58b4b9a22f

**Review by michalpaszkowski — APPROVED** [review]
Thank you @v-klochkov! LGTM!

**Comment by michalpaszkowski** [comment]
@v-klochkov Reverting the change due to [a failure](https://lab.llvm.org/buildbot/#/builders/55/builds/4171). This may require some investigation. Here is the [reverting pull request](https://github.com/llvm/llvm-project/pull/119657).

**Comment by v-klochkov** [comment]
> @v-klochkov Reverting the change due to [a failure](https://lab.llvm.org/buildbot/#/builders/55/builds/4171). This may require some investigation. Here is the [reverting pull request](https://github.com/llvm/llvm-project/pull/119657).

@michalpaszkowski - Thank you for the quick revert of that patch.
I cherry-picked this patch and added extra fix for the memory error: https://github.com/llvm/llvm-project/pull/121861
Please see: https://github.com/llvm/llvm-project/pull/121861#issuecomment-257417626

The error is reproducible only with memory sanitizer. 
Is there a way to run one of post-commit CI bots before the commit there?

**Comment by akshayrdeodhar** [comment]
@v-klochkov - I'm curious- just like for any fixed depth, we will be able to construct a pathological case where multiple equivalence classes will be created for the same base, couldn't we construct pathological cases for MergeEquivalenceClasses where the ladder of depth-1 jumps does not suffice either?

**Comment by v-klochkov** [comment]
> @v-klochkov - I'm curious- just like for any fixed depth, we will be able to construct a pathological case where multiple equivalence classes will be created for the same base, couldn't we construct pathological cases for MergeEquivalenceClasses where the ladder of depth-1 jumps does not suffice either?

Not sure I understand the question. The patch resolved the issue that came from the fact of having limited/fixed depth of the underlying-object (UO) search in the beginning/creating the eq classes. If the found UOs differ by 1 level  (e.g. UO2 is GEP of UO1), then MergeEquivalenceClasses should merge them and increase chances for vectorization. If it the closest distance between two UOs is 2 steps (GEP-GEP), they will not be merged. IMO, that is possible, but unlikely scenario.

@akshayrdeodhar - do you have some test or workload that can be vertorized, but suffers from 2+ diff between originally found UOs?

**Comment by nikic** [comment]
I'm somewhat confused by this change. Why add this kind of complex post-processing step instead of increasing the getUnderlyingObject() limits?

In fact, I did raise it from 6 to 10 recently (https://github.com/llvm/llvm-project/pull/143714). Is it possible that this functionality is now no longer needed?

**Pull request merged**
